src/main/java/com/avispl/symphony/dal/communicator/middleatlantic/MiddleAtlanticPowerUnitCommunicator.java

```diff
@@ -78,14 +82,46 @@ protected void authenticate() throws Exception {
                         .forEach(CompletableFuture::join), executor)
                 .get(30, TimeUnit.SECONDS);
 
-        executor.shutdown();
+        executor.shutdownNow();
 
         extendedStatistics.setStatistics(statistics);
         extendedStatistics.setControl(control);
-
         return new ArrayList<>(Arrays.asList(extendedStatistics));
     }
 
+    /**
+     * @return -1 or timeout if host is not reachable within
+     * the pingTimeout, a ping time in milliseconds otherwise
+     * if ping is 0ms it's rounded up to 1ms to avoid IU issues on Symphony portal
+     * @throws IOException
+     */
+    @Override
+    public int ping() throws IOException {
+        InetAddress inetAddress = InetAddress.getByName(this.getHost());
+        long totalTime = 0L;
+
+        for (int i = 0; i < this.getPingAttempts(); i++) {
+            long startTime = System.currentTimeMillis();
+            boolean reachable = inetAddress.isReachable(this.getPingTimeout());
+            long endTime = System.currentTimeMillis();
+
+            if (reachable) {
+                totalTime += (endTime - startTime);
+                if (this.logger.isTraceEnabled()) {
+                    this.logger.trace(String.format("OK: TCP ping attempt for host %s: success in %sms", this.getHost(), (endTime - startTime)));
+                }
+            } else {
+                int notReachedIn = Math.toIntExact(endTime - startTime);
+                if (this.logger.isDebugEnabled()) {
+                    this.logger.debug(String.format("TIMEOUT: TCP ping attempt for host %s: failed in %sms", this.getHost(), notReachedIn));
+                }
+                return notReachedIn < getPingTimeout() ? -1 : notReachedIn;
```

Comment:
Not sure that it is good to have two returns in this method. 
Is it ok to return positive value if the address was not reached?

Reply:
It's implemented like that to somewhat match the BaseDevice logic i.e the ping is timed out -> the ping result is returned right away. The only difference is that it's hard to tell why it was not reached exactly - it's either a timeout or something different, for that purpose the getPingTimeout() check was added, so it assumes that the ip address is not accessable (thus the ping is less than the timeout value) or it's a timeout+ value, so this value is returned notifying the ping() method consumer that the timeout value is reached